### PR TITLE
feat(ui-hydro): add correlation form

### LINKED ISSUE #1358: [ANT-652] Add "Spatial correlation" form
## Problem Statement

The tab "Spatial correlaton" must be added as on the old interface, see the screenshot below

image

# Spatial Correlation Form for Hydro

## User Story:

As a User using Antares Web, I want to input correlation coefficients between two areas within the Hydro section, so that I can define relationships between areas for hydro power simulations and enhance the accuracy of my study results.

## Acceptance Criteria:

1. The Spatial Correlation Form should be integrated within the Hydro section of the Antares Simulator web application.
2. A new "Spatial Correlation" tab should be added inside the Hydro section.
3. A separate form should be available for each area of the study, allowing users to input correlation coefficients between the current area and other areas.
4. A matrix visualization should be accessible to users, allowing them to view all the correlations between all the areas of a study.
5. The form should include two input fields labeled "Area 1" and "Area 2" for users to enter the names of the two areas they want to define a correlation for.
6. The form should include a third input field labeled "Coefficient" for users to enter the correlation coefficient as a float.

> An example of the expected input format should be: "area1%area2=coefficient."

7. A "Submit" button should be added to save the input data and send it to the FastAPI backend.
8. Input validation should be implemented to ensure that:

> Area names (Area 1 and Area 2) are non-empty strings.
> The correlation coefficient is a float, a valid number between -100 and 100, inclusive.
> Error messages are shown if input fields do not meet the above requirements.

9. The user interface of the Spatial Correlation Form should be visually consistent with the rest of application.
10. The user's input should be correctly formatted and sent to the FastAPI backend.
11. The FastAPI backend should have endpoints for getting spatial correlations, updating spatial correlations, and getting the overall matrix.
12. The FastAPI backend should successfully update the ini file that stores all the spatial correlations of a study.
13. The new feature should not negatively impact the application's overall performance.
14. Clear and concise error messages should be displayed to the user if there is any issue with updating the ini file.

> Show a success message or visual indicator when a user successfully submits a new correlation coefficient or updates an existing one, providing users with feedback on their actions.

15. The user should be able to view the current spatial correlations of a study in the Spatial Correlation Form before making changes.
16. The frontend should display the area-specific forms, the overall matrix, and any information retrieved from the backend endpoints for spatial correlations.
17. The implementation should be tested to ensure proper functionality and a smooth user experience.

### PR DESCRIPTION
**Pull request by hdinia**
fix #1358

### CODE REVIEW

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Allocation/index.tsx [hunk 1]
```diff
@@ -23,7 +23,7 @@ function Allocation() {
   ////////////////////////////////////////////////////////////////
 
   const handleSubmit = (data: SubmitHandlerPlus<AllocationFormFields>) => {
-    return setAllocationFormFields(studyId, areaId, {
+    setAllocationFormFields(studyId, areaId, {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Allocation/index.tsx]: Keep the return is necessary

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Allocation/Fields.tsx [hunk 2]
```diff
@@ -1,70 +1,52 @@
...
+          key={item.id}
+          field={item}
+          index={index}
+          label={areasById?.[item.areaId]?.name}
+          remove={remove}
+          fieldsLength={fields.length}
+        />
+      )}
+      options={options}
+      onAdd={(value: string) =>
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Allocation/Fields.tsx]: Remove type cast (fix it in DynamicList if necessary)

File: webapp/src/components/common/DynamicList.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  ////////////////////////////////////////////////////////////////
+  // JSX
+  ////////////////////////////////////////////////////////////////
+
+  return (
+    <Grid container direction="column" spacing={2}>
+      <Grid item>
+        {items.map((item, index) => (
+          <React.Fragment key={item.id}>
+            {renderItem(item, index)}
```
Comment (skamril) [inline webapp/src/components/common/DynamicList.tsx]: The delete button after the render item. Add `onDelete` prop.

File: webapp/src/redux/selectors.ts [hunk 4]
```diff
@@ -214,6 +214,19 @@ export const getAreas = createSelector(getStudySynthesis, (synthesis) => {
   return [];
 });
 
+export const getAreasById = createSelector(getStudySynthesis, (synthesis) => {
+  if (synthesis) {
+    return Object.keys(synthesis.areas).reduce(
+      (acc, id) => ({
+        ...acc,
+        [id]: { ...synthesis.areas[id], id },
```
Comment (skamril) [inline webapp/src/redux/selectors.ts]: It's more performant to mutate in a reduce, it's not necessary to create a new object:

```
acc[id] = ...
return acc
```

File: webapp/src/redux/selectors.ts [hunk 5]
```diff
@@ -214,6 +214,19 @@ export const getAreas = createSelector(getStudySynthesis, (synthesis) => {
...
+      }),
+      {}
+    ) as Record<string, Area & { id: string }>;
```
Comment (skamril) [inline webapp/src/redux/selectors.ts]: `{} as Record<string, Area & { id: string }>;`

File: webapp/src/components/common/DynamicList.tsx [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  onAdd: (value: string) => void;
+}
+
+function DynamicList<T extends { id: string }>({
+  items,
+  renderItem,
+  options,
+  onAdd,
+}: DynamicListProps<T>) {
+  ////////////////////////////////////////////////////////////////
```
Comment (skamril) [inline webapp/src/components/common/DynamicList.tsx]: Remove block

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/utils.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import { StudyMetadata, Area } from "../../../../../../../../common/types";
+import client from "../../../../../../../../services/api/client";
+
+////////////////////////////////////////////////////////////////
+// Types
+////////////////////////////////////////////////////////////////
+
+export interface CorrelationField {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/utils.ts]: Duplicate with AllocationField

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts [hunk 8]
```diff
@@ -156,3 +165,31 @@ function generateColumns(columnSuffix = ""): string[] {
...
+
+/**
+ * A custom hook that returns area options filtered by the given fields.
+ *
+ * @param fields - An array of objects containing areaId and coefficient.
+ * @returns A filtered list of AreaOptions.
+ */
+export function useAreasOptions(
+  fields: { areaId: string; coefficient: number }[]
+): AreaOption[] {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts]: `: DynamicListProps["options"]`

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts [hunk 9]
```diff
@@ -34,6 +38,11 @@ export interface HydroRoute {
   type: number;
 }
 
+interface AreaOption {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts]: Remove

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts [hunk 10]
```diff
@@ -156,3 +165,31 @@ function generateColumns(columnSuffix = ""): string[] {
   }
   return columns;
 }
+
+/**
+ * A custom hook that returns area options filtered by the given fields.
+ *
+ * @param fields - An array of objects containing areaId and coefficient.
+ * @returns A filtered list of AreaOptions.
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts]: "A filtered list of DynamicList's options."

But we already see the return type in TS...

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts [hunk 11]
```diff
@@ -156,3 +165,31 @@ function generateColumns(columnSuffix = ""): string[] {
   }
   return columns;
 }
+
+/**
+ * A custom hook that returns area options filtered by the given fields.
+ *
+ * @param fields - An array of objects containing areaId and coefficient.
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts]: If you want to use JSDoc...

```
 * @param {Object[]} fields ...
 * @param {string} fields[].areaId ...
 * @param {number} fields[].coefficient ...
```

But in my option, in TS there is no necessity to describe arguments

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts [hunk 12]
```diff
@@ -156,3 +165,31 @@ function generateColumns(columnSuffix = ""): string[] {
...
+ * @returns A filtered list of AreaOptions.
+ */
+export function useAreasOptions(
+  fields: { areaId: string; coefficient: number }[]
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/utils.ts]: Create a type for both allocation and correlation, then use it here

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/index.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    study: { id: studyId },
+  } = useOutletContext<{ study: StudyMetadata }>();
+  const areaId = useAppSelector(getCurrentAreaId);
+
+  ////////////////////////////////////////////////////////////////
+  // Event handlers
+  ////////////////////////////////////////////////////////////////
+
+  const handleSubmit = (data: SubmitHandlerPlus<CorrelationFormFields>) => {
+    setCorrelationFormFields(studyId, areaId, {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/index.tsx]: Return the promise

**Review by skamril — CHANGES_REQUESTED** [review]

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+interface Props {
+  field: FieldArrayWithId<CorrelationFormFields, "correlation">;
+  index: number;
+  label: string;
+}
+
+function CorrelationField({ field, index, label }: Props) {
+  const { control } = useFormContextPlus<CorrelationFormFields>();
+
+  ////////////////////////////////////////////////////////////////
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx:19]: Remove block

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      <Grid item xs={4} md={2}>
+        <NumberFE
+          key={field.id}
+          name={`correlation.${index}.coefficient`}
+          size="small"
+          control={control}
+          rules={{
+            min: {
+              value: -100,
+              message: "min: -100%",
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx]: Use `form.field.minValue`

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+            },
+            max: {
+              value: 100,
+              message: "max: 100%",
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx]: Add `form.field.maxValue`

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx [hunk 17]
```diff
@@ -0,0 +1,55 @@
+import { Typography, Grid } from "@mui/material";
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/CorrelationField.tsx:1]: Create AreaCoefFE (only `names` and rules `change`)

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/utils.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
+import { StudyMetadata, Area } from "../../../../../../../../common/types";
+import client from "../../../../../../../../services/api/client";
+import { AreaCoefficientField } from "../utils";
+
+////////////////////////////////////////////////////////////////
+// Types
+////////////////////////////////////////////////////////////////
+
+export interface CorrelationFormFields {
+  correlation: AreaCoefficientField[];
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/utils.ts]: Like I said to you. AreaCoefficientField type it is not a field but an item of the array.
AreaCoefficientField[] => array of fields ❌ 
So rename it (AreaCoefObj for example)

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/hooks/useAreasOptions.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
+import { useMemo } from "react";
+import { useOutletContext } from "react-router";
+import { StudyMetadata } from "../../../../../../../../common/types";
+import useAppSelector from "../../../../../../../../redux/hooks/useAppSelector";
+import { getAreas } from "../../../../../../../../redux/selectors";
+import { DynamicListProps } from "../../../../../../../common/DynamicList";
+import { AreaCoefficientField } from "../utils";
+
+export function useAreasOptions(
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/hooks/useAreasOptions.ts:9]: `useAreaOptions`, without "s" to area.
Don't forget to rename the file.

File: webapp/src/components/common/DynamicList.tsx [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+  value: string;
+}
+
+export interface DynamicListProps<T extends { id: string } = { id: string }> {
+  items: T[];
+  renderItem: (item: T, index: number) => React.ReactNode;
+  options: ListOption[];
+  onAdd: (value: string) => void;
+  onRemove: (index: number) => void;
+  disableRemove?: boolean;
```
Comment (skamril) [inline webapp/src/components/common/DynamicList.tsx]: Better `allowEmpty?: boolean`, `true` by default

**Review by skamril — CHANGES_REQUESTED** [review]

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/hooks/useAreasOptions.ts (hunk 19 above)
Comment (hdinia) [reply webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/hooks/useAreasOptions.ts:9]: `useAreaOptions`  This name suggests that the hook provides options for a single area.
`useAreasOptions` This name implies that the hook provides options for multiple areas.

File: webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/utils.ts (hunk 18 above)
Comment (hdinia) [reply webapp/src/components/App/Singlestudy/explore/Modelization/Areas/Hydro/Correlation/utils.ts]: I believe that the name `AreaCoefficientField` is more descriptive and better represents the purpose of the object within the form,as it concisely represents separate fields within the correlation form, with distinct area IDs and coefficients, compared to a more generic name like AreaCoefObj.

**Review by skamril — APPROVED** [review]

**Pull request merged**